server/src/rate_limit/mod.rs

```diff
@@ -2,17 +2,182 @@ pub mod rate_limiter;
 
 use super::{IPAddr, Settings};
 use crate::api::APIError;
+use crate::settings::RateLimitConfig;
+use actix_web::dev::{Service, ServiceRequest, ServiceResponse, Transform};
 use failure::Error;
+use futures::future::{ok, Ready};
 use log::warn;
-use rate_limiter::RateLimiter;
+use rate_limiter::{RateLimitType, RateLimiter};
 use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
 use std::sync::Arc;
-use std::sync::Mutex;
+use std::task::{Context, Poll};
 use std::time::SystemTime;
 use strum::IntoEnumIterator;
+use tokio::sync::Mutex;
 
 #[derive(Debug, Clone)]
-pub struct RateLimitInfo {
-  pub rate_limiter: Arc<Mutex<RateLimiter>>,
-  pub ip: IPAddr,
+pub struct RateLimit(pub Arc<Mutex<RateLimiter>>);
+
+#[derive(Debug, Clone)]
+pub struct RateLimited(Arc<Mutex<RateLimiter>>, RateLimitType);
+
+pub struct RateLimitedMiddleware<S>(RateLimited, S);
```

Comment:
This is personal preference, but could you change these to full structs? Unit / Tuple structs are a big peeve of mine, I can't stand seeing `.0` and `.1`s in code, then having to trace down what the thing actually is. I know its more verbose, but a lot easier for me to understand.

So these could be:

```
pub struct RateLimit {
  pub rate_limiter: Arc<Mutex<RateLimiter>>,
}

pub struct RateLimited {
  rate_limiter: Arc<Mutex<RateLimiter>>,
  type_: RateLimitType,
}

pub struct RateLimitedMiddleware<S> {
  rate_limited: RateLimited, 
  whatever: S
}

Reply:
Agreed, tuples in Rust are really bad for readability.